### PR TITLE
CRM-19060 fix Payment table join in membership Detail report to ensur…

### PR DESCRIPTION
**Pull request by seamuslee001**
…e latest membership payment is picked up

### CODE REVIEW

**Comment by seamuslee001** [comment]
@eileenmcnaughton @sqweets

This is a re-do of PR https://github.com/civicrm/civicrm-core/pull/8687/files I have attempted to improve it by getting the financial type from the membership_type and also using civicrm_line_item rather than civicrm_membership_payment. Still not 100% sure if this is perfect but hopefully gets us closer. Also it makes me feel like we should drop the financial_type filter from the ui as well as i'm not sure what the intent of that would be

**Comment by sqweets** [comment]
@seamuslee001 @eileenmcnaughton I tested this patch and found that on line 330 `$types = $this->_params['tid_value'];` when $types = 0 you get a syntax error for the WHERE clause on 347.   I defaulted the type to 2 (membership dues) and I found that the original issue wasn't fixed.  It was pulling in the first membership payment and not the latest.

**Comment by seamuslee001** [comment]
@sqweets Ellen thanks for that I think i have fixed for when no membership types are set and the latest payment issue. Also i have ensured that if a date filter on receive date is used that it is passed through to the find latest payment sql. Can you re test again.

File: CRM/Report/Form/Member/Detail.php [hunk 1]
```diff
@@ -326,13 +326,59 @@ public function from() {
...
+      if (isset($params['receive_date_relative']) || isset($params['receive_date_from']) || isset($params['receive_date_to'])) {
+        $dateFilter = parent::dateClause('receive_date', $params['receive_date_relative'],
+          $params['receive_date_from'], $params['receive_date_to']);
+      }
+      else {
+        $dateFilter = FALSE;
+      }
+      if (is_array($types) && !empty($types)) {
+        foreach ($types as $membershipType) {
+          $result = CiviCRM_Api3('membershipType', 'get', array('id' => $membershipType));
```
Comment (eileenmcnaughton) [inline CRM/Report/Form/Member/Detail.php]: your capitalisation is wrong civicrm_api3

**Comment by eileenmcnaughton** [comment]
Have we tested this on any larger sites? The subquery may be less performant

**Comment by seamuslee001** [comment]
@eileenmcnaughton thanks for that, I have just tested on AUG sandbox and report seems to be reasonable in terms of performance, it wasn't with all options selected or heaps of complex filtering but i think its ok

**Comment by sqweets** [comment]
My testing looks good, thanks very much for getting this fixed.

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by davejenx** [comment]
I've tested this by applying the PR as a patch to a 4.6.21 test site with 83k memberships (didn't have a suitable 4.7 available), on a lowish-spec test server. I found that with the patch, the report query was orders of magnitude slower and more importantly gave different results.

E.g. running member detail report with default criteria (groupings: membership, contrib status: completed):

Without patch:
results: 72058
execute query time: 6.5s

With patch:
results: 64889
execute query time: 811s

I cleared the query cache before each run and timed multiple runs.

I've looked at a couple of examples of contacts who appeared in the results without the patch but not with it.

- The first has a membership paid by monthly direct debit, using the Smart Debit extension. The line items for this contact all have entity_table = 'civicrm_contribution', so none were found with 'civicrm_membership'. That may be a weakness in the extension.

- The second has 3 line items with entity_table = 'civicrm_membership'. I looked into what the report queries were doing. When populating the temporary table in contributionTableJoin():

      CREATE TEMPORARY TABLE civicrm_report_memcontr_temp_a3f7b977b7ef4c89217fe9399f7d71b8 (index (contact_id, receive_date, entity_id))
      SELECT cc.contact_id, max(cc.receive_date) as receive_date, cli.entity_id
      FROM civicrm_line_item cli
      INNER JOIN civicrm_contribution cc
      ON cc.id = cli.contribution_id
      WHERE cli.financial_type_id IN (2)
      AND cli.entity_table = 'civicrm_membership'
      GROUP BY cc.contact_id

I think the condition on financial_type_id is the problem there. The line items for the second contact have other financial_type_ids. It looks as though a different financial type may have been in use for some membership types at the time, whereas contributionTableJoin() checks current membership types' financial_type_ids.

I discussed with @eileenmcnaughton at the Edale sprint. Knotty problems like this are likely to result in inconsistencies particularly affecting historical data. Personally I find civicrm_membership_payment a more reliable record of membership payments; I understand that it is to be deprecated but I think doing so will require a well-tested procedure to reconcile data against the more reliable source for historical data, civicrm_membership_payment.

**Comment by seamuslee001** [comment]
thanks @davejenx I might try going back to civicrm_membership_payment in that case

**Comment by eileenmcnaughton** [comment]
@davejenx just trying to get this old PR to a resolution - @seamuslee001 has actually made changes to address your points - wondering if you could re-test

**Comment by eileenmcnaughton** [comment]
@davejenx did you spot this?

**Comment by davejenx** [comment]
@eileenmcnaughton Yes, sorry, I barely got to open a text editor last week, hope to make some time for it this week.

**Comment by eileenmcnaughton** [comment]
@davejenx thanks - good to know it's on your radar

**Comment by davejenx** [comment]
@eileenmcnaughton @seamuslee001 Sorry for the excessive delay! Re-testing on a recently upgraded 4.7.23 dev copy of the 4.6 site that I tested on last year. On a different server.

### Scenario 1 - Testing with default fields, filters etc - unconstrained search.

Times are for the executeQuery in buildRows().
RESET QUERY CACHE between trials.

**Results without patch:**
94806 records, average of fastest 3 trials: 2.8s

**Results with patch:**
94806 records, average of fastest 3 trials: 1578s

So same number of records but the modified report query was over 500x slower.

I than made a change to the indexes on the temporary table, which speeded things up a lot.

```
-      CREATE TEMPORARY TABLE $tempTable (index (contact_id, receive_date, entity_id))
+      CREATE TEMPORARY TABLE $tempTable (INDEX (contact_id), INDEX (receive_date), INDEX (entity_id))
```

**Results with patch + single-column indexes:**
94806 records, average of fastest 3 trials: 2.4s

To make sure results are consistent, I tried a search with more filters:

### Scenario 2 - Specified membership type, receive date & group.

**Results without patch:**
1482 records, average of fastest 3 trials: 0.11s

**Results with patch:**
1482 records, average of fastest 3 trials: 0.06s

**Results with patch + my index changes:**
1482 records, average of fastest 3 trials: 0.06s

Interesting! With these filters, the query is equally fast with or without the index changes.

Presumably when the query involves receive_date and contact_id (e.g. if group is specified), MySQL makes use of the multiple-column index on (contact_id, receive_date, entity_id). But when those fields aren't involved, it can't use that index and there is no index on entity_id alone. Maybe a multiple-column index could work if the order of columns is changed, given that:

> MySQL can use multiple-column indexes for queries that test all the columns in the index, or queries that test just the first column, the first two columns, the first three columns, and so on.
- https://dev.mysql.com/doc/refman/5.7/en/multiple-column-indexes.html

I tried:

```
-      CREATE TEMPORARY TABLE $tempTable (index (contact_id, receive_date, entity_id))
+      CREATE TEMPORARY TABLE $tempTable (INDEX (entity_id, contact_id, receive_date))
```

Scenario 1 (unconstrained search):

**Results with patch + changed multiple-column index:**
94806 records, average of fastest 3 trials: 2.4s

Scenario 2 (specified membership type, receive date & group):

**Results with patch + changed multiple-column index:**
1482 records, average of fastest 3 trials: 0.06s

Also tried leaving group out of the criteria, to see if the changed multiple-column index could still be used:
Scenario 3 (specified membership type & receive date):

**Results without patch:**
2716 records, average of fastest 3 trials: 0.16s

**Results with patch:**
2715 records, average of fastest 3 trials: 0.07s

**Results with patch + changed multiple-column index:**
2715 records, average of fastest 3 trials: 0.07s

Annoyingly there's a discrepancy of 1 in the number of records for Scenario 3. I've identified the contact who's excluded with the patch. She has a membership of the specified type. There are 3 membership_payment records for this membership. The 3 linked contributions are all recurring, with dates 1st Feb 2015, 1st Feb 2016 & 1st Feb 2017. I had filtered on receive date in 2017, so the last should be included.

I tried manually populating the temporary table & seeing what was there for this contact. max(receive_date) came out as 1st May 2017. Drilling down, this was the max receive_date of 6 contributions, dated 1st Jan 2017 to 1st May 2017. Only 1 of those 6 was for the contact's one existing membership though: the one dated 1st Feb 2017. The other 5 were for another membership. It looks like this contact made some contributions that were paying for another contact's membership. Due to the differences in the queries, the patched report is including those when looking for the most recent contribution, so in this case is producing a mismatch on receive_date.

Have to go now, away for a few days. To summarise, the results are looking good with the temp table index changed to either the 3 separate indexes or the reordered multiple-field index. There's an edge case where the results differ slightly from the original report. I can look at that a bit more if needed after I get back.

**Comment by seamuslee001** [comment]
@davejenx Thank you so much for that testing and write up, really really useful. I think i may have a fix for the multiple membership issue thing. I think i'll go with the 3 column combo index. That is a really useful bit of MySQL documentation as well

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by eileenmcnaughton** [comment]
Thanks @davejenx  for all this extra review. Let me know if you think @seamuslee001 last changes are OK - you don't need to test as thoroughly this time - as long as they are not worse than when you last tested this seems mergable

**Comment by eileenmcnaughton** [comment]
@davejenx just waiting on your re-confirmation...

**Comment by monishdeb** [comment]
I have tested the Membership Detail report on top of the changes, in my local, it works for me. Also from code pov it's good to merge.

**Comment by monishdeb** [comment]
@davejenx @eileenmcnaughton can we merge this?

**Comment by eileenmcnaughton** [comment]
I think if you have retested after the last changes that seems OK - @davejenx hopefully you can give master or the next rc a test

**Comment by davejenx** [comment]
Hmm, I'm now getting slower results. Now testing against 4.7.26RC (+ this patch), previously 4.7.23 .

For Scenario 1 - Testing with default fields, filters etc - unconstrained search:
Not as slow as an earlier version but still up from < 3s for unpatched to 330s, so more than 100x slower than the original version without temp tables.

For Scenario 2 - Specified membership type, receive date & group:
up from 0.1s unpatched to 4.5s, so 45x slower than the original.

After many re-tests, including trying some of the previous versions of the patch, it's looking like all the variants involving the temporary table are slow - but the current variant, with the compound index re-ordered, is not as slow as the original ordering.

I think I must have been led astray in previous testing by Opcache holding onto a previous version of Detail.php . I was religiously clearing the query cache but think I overlooked Opcache. Either that or something else in 4.7.26RC vs 4.7.23 is affecting it, but I don't see how given that the report queries are self-contained. Re-testing with a php-fpm reload after each swap of Detail.php, I'm consistently getting about 330s for scenario 1 and 4.5s for scenario 2 with the current temp table variant (or the one with 3 separate indexes) - compared to about 2.5s and 0.1s for the original with no temp table.

The first temp table variant, with index (contact_id, receive_date, entity_id), is slower still: about > 1800s (gateway timeout) for scenario 1, 4.6s for scenario 2.

For Scenario 3 (specified membership type & receive date): the current version fixes the discrepancy in the number of records returned vs the original version. Time for this is is 8.8s compared to 0.15s for the unpatched version.

**Comment by agileware-justin** [comment]
https://lab.civicrm.org/dev/core/-/issues/2374

**Pull request merged**